data-prepper-plugins/opensearch/src/main/java/com/amazon/dataprepper/plugins/sink/opensearch/OpenSearchSink.java

```diff
@@ -123,47 +127,91 @@ public void doOutput(final Collection<Record<Object>> records) {
       return;
     }
 
+    AccumulatingBulkRequest<BulkOperation, BulkRequest> bulkRequest = bulkRequestSupplier.get();
 
+    if(IndexConfiguration.BULK_ACTION_CREATE.equals(action)) {
 
-    AccumulatingBulkRequest<BulkOperation, BulkRequest> bulkRequest = bulkRequestSupplier.get();
-    for (final Record<Object> record : records) {
-      final SerializedJson document = getDocument(record.getData());
+      for (final Record<Object> record : records) {
+        final SerializedJson document = getDocument(record.getData());
+
+        final CreateOperation.Builder<Object> createOperationsBuilder = new CreateOperation.Builder<>()
+                .index(indexManager.getIndexAlias())
+                .document(document);
 
-      final IndexOperation.Builder<Object> indexOperationBuilder = new IndexOperation.Builder<>()
-              .index(indexManager.getIndexAlias())
-              .document(document);
+        final Map documentAsMap;
+        try {
+          documentAsMap = objectMapper.readValue(document.getSerializedJson(), Map.class);
+        } catch (IOException e) {
+          throw new RuntimeException(e);
+        }
+        if (documentAsMap != null) {
+          final String docId = (String) documentAsMap.get(documentIdField);
+          if (docId != null) {
+            createOperationsBuilder.id(docId);
+          }
+        }
 
+        final BulkOperation createBulkOperation = new BulkOperation.Builder()
+                .create(createOperationsBuilder.build())
+                .build();
 
-      final Map documentAsMap;
-      try {
-        documentAsMap = objectMapper.readValue(document.getSerializedJson(), Map.class);
-      } catch (IOException e) {
-        throw new RuntimeException(e);
+        final long estimatedBytesBeforeAdd = bulkRequest.estimateSizeInBytesWithDocument(createBulkOperation);
+        if (bulkSize >= 0 && estimatedBytesBeforeAdd >= bulkSize && bulkRequest.getOperationsCount() > 0) {
+          flushBatch(bulkRequest);
+          bulkRequest = bulkRequestSupplier.get();
+        }
+        bulkRequest.addOperation(createBulkOperation);
       }
-      if(documentAsMap != null) {
-        final String docId = (String) documentAsMap.get(documentIdField);
-        if (docId != null) {
-          indexOperationBuilder.id(docId);
+
+      // Flush the remaining requests
+      if (bulkRequest.getOperationsCount() > 0) {
+        flushBatch(bulkRequest);
+      }
+
+    } else if(IndexConfiguration.BULK_ACTION_INDEX.equals(action)) {
```

Comment:
We can limit the amount of duplicate code by refactoring the `if create, else if index` to look something like this. It would also be nice to move creation of the `BulkOperation` to a Factory class with functions of `provideBulkOperationForCreate` and `provideBulkOperationForIndex`, but that isn't necessary for this PR.

```java
for (final Record<Object> record : records) {
        final SerializedJson document = getDocument(record.getData());

        final Optional<String> docId = getDocumentIdFromDocument(document);

        BulkOperation bulkOperation;
        if (action.equals("index")) {
          final IndexOperation.Builder<Object> indexOperationBuilder = new IndexOperation.Builder<>()
            .index(indexManager.getIndexAlias());
            .document(document);

          if (docId.isPresent()) {
            indexOperationBuilder.id(docId);
          }

          indexOperationBuilder.build();

          bulkOperation = new BulkOperation.Builder()
                    .index(indexOperationBuilder.build())
                    .build();

        } else if (action.equals("create")) {

          final CreateOperation.Builder<Object> createOperationBuilder = new CreateOperation.Builder<>()
            .index(indexManager.getIndexAlias());
            .document(document);

          if (docId.isPresent()) {
            createOperationBuilder.id(docId);
          }

          createOperationBuilder.build();

          bulkOperation = new BulkOperation.Builder()
                .create(createOperationsBuilder.build())
                .build();
        }
                

        final long estimatedBytesBeforeAdd = bulkRequest.estimateSizeInBytesWithDocument(bulkOperation);
        if (bulkSize >= 0 && estimatedBytesBeforeAdd >= bulkSize && bulkRequest.getOperationsCount() > 0) {
          flushBatch(bulkRequest);
          bulkRequest = bulkRequestSupplier.get();
        }
        bulkRequest.addOperation(bulkOperation);
      }

      // Flush the remaining requests
      if (bulkRequest.getOperationsCount() > 0) {
        flushBatch(bulkRequest);
      }

    }
}

private Optional<String> getDocumentIdFromDocument(final SerializedJson document) {
  final Map documentAsMap;
        try {
          documentAsMap = objectMapper.readValue(document.getSerializedJson(), Map.class);
        } catch (IOException e) {
          throw new RuntimeException(e);
        }
        if (documentAsMap != null) {
          final String docId = (String) documentAsMap.get(documentIdField);
          if (docId != null) {
            return Optional.of(docId);
          }
        }
  return Optional.empty();
}
```